test/replication-luatest/election_fencing_test.lua

```diff
@@ -0,0 +1,288 @@
+local luatest = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+local g_async = luatest.group('fencing_async', {
+    {election_mode = 'manual'}, {election_mode = 'candidate'}})
+local g_sync = luatest.group('fencing_sync')
+
+local SHORT_TIMEOUT = 0.1
+local LONG_TIMEOUT = 1000
+local DEATH_TIMEOUT = 2 * SHORT_TIMEOUT
+
+local function promote(server)
+    luatest.helpers.retrying({}, function()
+        luatest.assert(server:exec(function()
+            pcall(box.ctl.promote)
+            if box.info.election.state ~= 'leader' then
+                return false
+            end
+            if box.info.synchro.queue.owner ~= box.info.id then
+                return false
+            end
+            return true
+        end))
+    end)
+    server:exec(function() box.ctl.wait_rw() end)
+end
+
+local function wait_sync(leader, servers)
+    local vclock = leader:get_vclock()
+    vclock[0] = nil
+    for _, server in ipairs(servers) do
+        server:wait_vclock(vclock)
+    end
+end
+
+local function wait_disconnected(node_1, node_2)
+    luatest.helpers.retrying({}, function()
+        luatest.assert(node_1:exec(function(i)
+            return (box.info.replication[i].downstream.status == 'stopped' or
+                    box.info.replication[i].upstream == nil or
+                    box.info.replication[i].upstream.status == 'disconnected')
```

Comment:
`box.info` is expensive, I would commend to save `box.info.replication[i]` into a variable and use it in these expressions. Also you don't need `()` around the statement in `return`. Same in `wait_connected()`.

Reply:
Fixed